### ISSUE TITLE
Questions regarding further devops improvements

### ISSUE THREAD

**Issue by ggravlingen** [Arelle/Arelle#268]
In other Python projects I'm working in, [pre-commit hooks ](https://pre-commit.com/) have been setup to help avoid commiting code that will not pass the CI/CD for quality reasons. Is this something you have considered doing? In that case, I'm happy to set it up.

Also, have you also considered [running Black](https://black.readthedocs.io/en/stable/) on the code?

**Comment by sagesmith-wf** [comment]
It'll ultimately up to @austinmatherne-wk, but both of these seem pretty interesting to me at least. Black combined with some sort of commit hook might be interesting though it will likely conflait a bunch of changes at the beginning (which I am at least alright with). In general the `pre-commit` hooks haven't been that interesting in the past but I'm open to the idea of trying them with an open source repo

**Comment by ggravlingen** [comment]
The way I've seen it done (and I've successfully done it myself) is to run Black once on the entire codebase to set and merge a baseline, and then add a check to the CI/CD to make sure all new commits follow the guidelines. The pre-commit will be of great help on a commit-by-commit basis.

**Comment by sagesmith-wf** [comment]
> run Black once on the entire codebase to set and merge a baseline

👍 that seems like a decent idea, though its unfortunate that it will likely make git blame a bit unreliable but it should still be available in history ultimately so again, not too concerned.

**Comment by ggravlingen** [comment]
> > run Black once on the entire codebase to set and merge a baseline
> 
> 
> 
> 👍 that seems like a decent idea, though its unfortunate that it will likely make git blame a bit unreliable but it should still be available in history ultimately so again, not too concerned.

I haven't tried this myself but it sounds like something that could be useful in this context: https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html

**Comment by austinmatherne-wk** [comment]
> In other Python projects I'm working in, [pre-commit hooks ](https://pre-commit.com/) have been setup to help avoid commiting code that will not pass the CI/CD for quality reasons. Is this something you have considered doing? In that case, I'm happy to set it up.
> 
> Also, have you also considered [running Black](https://black.readthedocs.io/en/stable/) on the code?

Yes to both questions. With CI we also want to run [bandit](https://github.com/PyCQA/bandit) and linting (more problematic because arelle doesn't follow standard style conventions).

I did a migration on a large typescript repo a while back using [prettier](https://prettier.io/) and pre-commit hooks. It went well, but arelle is more complicated. There are private forks (SEC) that diverge and aren't frequently merged with the open source master branch. We don't want to cause many merge conflicts for those folks. Any mass formatting updates should only happen when the forks are already relatively in sync and we've properly warned folks that changes are coming.

This is on our radar and will be prioritized once we've finished replacing the manual builds uploaded to arelle.org with reproducible CI artifacts and have a proper package published to PyPI.
